File: 005_rust_ducks/src/noise_strategy.rs
use crate::duck_traits::Duck;

pub struct SilentStrategy;
pub struct QuackStrategy;

pub trait NoiseStrategy {
    fn sound<T: Duck>(t: &T) -> &str;
}

impl NoiseStrategy for SilentStrategy {
    fn sound<T: Duck>(_t: &T) -> &str {
        "no"
    }
}
impl NoiseStrategy for QuackStrategy {
    fn sound<T: Duck>(_t: &T) -> &str {
        "quack"
    }
}

File: 005_duck_factory/src/main/java/com/github/kgrech/ducks/strategy/NoiseStrategy.java
package com.github.kgrech.ducks.strategy;

import com.github.kgrech.ducks.Duck;

public enum NoiseStrategy {
    QuackStrategy {
        @Override
        public String sound(Duck duck) {
            return "quack";
        }
    },
    SilentStrategy {
        @Override
        public String sound(Duck duck) {
            return "no";
        }
    };

    public abstract String sound(Duck duck);
}
